### PR TITLE
Include source file hash in crate_hash.

### PR DESCRIPTION
**Pull request by cjgillot**
The current crate hash contains the hash of HIR nodes.
However, HIR does not necessarily contain all the information which transits in the query system (for instance stored in ResolverOutputs).
To avoid issues, this PR switches to using the hash of the source files instead.

r? @Aaron1011

### CODE REVIEW

**Comment by eddyb** [comment]
I wonder if there is some risk here of missing something that can cause HIR to differ.

Hard to check, would probably have to hack up incremental to store both hashes and assert that if the hash from this PR is unchanged, the older HIR-based hash is also unchanged (but that doesn't protect from anything weirder, and it's not like `rustc` has "phone home" functionality for freak accidents, either).

**Comment by petrochenkov** [comment]
>missing something that can cause HIR to differ

Something like environment variables, for example?
`env!("FOO")` always has the same source, but its result in HIR may change with time.

Also proc macros can produce random data out of thin air, I guess.

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 1]
```diff
@@ -1091,56 +1090,21 @@ impl<'hir> intravisit::Map<'hir> for Map<'hir> {
...
-        .map(|source_file| source_file.name_hash)
+        .map(|source_file| (source_file.name_hash, source_file.src_hash))
         .collect();
+    source_file_hashes.sort_unstable_by_key(|&(name_hash, _)| name_hash);
 
-    source_file_names.sort_unstable();
-
-    let mut hcx = tcx.create_stable_hashing_context();
-    let mut stable_hasher = StableHasher::new();
-    hir_body_hash.hash_stable(&mut hcx, &mut stable_hasher);
```
Comment (bjorn3) [inline compiler/rustc_middle/src/hir/map/mod.rs:1117]: The hir still needs to be hashed as proc macros don't need to be deterministic.

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 2]
```diff
@@ -1140,20 +1147,6 @@ pub(super) fn crate_hash(tcx: TyCtxt<'_>, crate_num: CrateNum) -> Svh {
     Svh::new(crate_hash.to_smaller_hash())
 }
 
-fn upstream_crates(tcx: TyCtxt<'_>) -> Vec<(StableCrateId, Svh)> {
```
Comment (Aaron1011) [inline compiler/rustc_middle/src/hir/map/mod.rs:1143]: Is there any particular reason for inlining this method? I don't have a strong preference about where this code should be, but it's currently adding unrelated changes to the diff.
